[Proposal] Extending Static Member Lookup in Generic Contexts (#1212)
apple/swift-evolution: proposals/xxxx-extend-generic-static-member-lookup.md

```swift
extension View {
  public func toggleStyle<S: ToggleStyle>(_ style: S) -> some View
}
```

which accepts instances of the `ToggleStyle` protocol, e.g.

```swift
public protocol ToggleStyle {
  associatedtype Body: View
  func makeBody(configuration: Configuration) -> some View
}

public struct DefaultToggleStyle: ToggleStyle { ... }
public struct SwitchToggleStyle: ToggleStyle { ... }
public struct CheckboxToggleStyle: ToggleStyle { ... }
```

Today, SwiftUI apps must write the full name of the concrete conformers to `ToggleStyle` when using the `toggleStyle` modifier:

```swift
Toggle("Wi-Fi", isOn: $isWiFiEnabled)
  .toggleStyle(SwitchToggleStyle())
```

However, this approach has a few downsides:

* **Repetitive:** Only the “Switch” component of the style name is important, since we already know that the modifier expects a type of `ToggleStyle`.
* **Poor discoverability:** There is no autocomplete support to expose the available `ToggleStyle` types to choose from, so you have to know them in advance.

> These downsides are impossible to avoid for generic parameters like above, which discourages generalizing functions. API designers should not have to chose between good design and easy-to-read code.

Reviewer: ```suggestion
These downsides are impossible to avoid for generic parameters like above, which discourages generalizing functions. API designers should not have to choose between good design and easy-to-read code.
```

* Do a global lookup for any type which conforms to the given protocol and use it as a base;
* Require that the result type of a member declaration conforms to the declaring protocol.

The second option is a much better choice that avoids having to do a global lookup and conformance checking and is consistent with semantics of leading dot syntax, namely, the requirement that result and base types of the chain have to be equivalent. This leads to a new rule: if the result type of a static member conforms to the declaring protocol, it should be possible to reference such a member on a protocol metatype, using leading dot syntax, by implicitly replacing the protocol with a conforming type.

> > **Note:** If member returns a function type or an optional value type-checker considers “result type” (for purposes of base type inference) to be a result type of a function type and/or wrapped value of an optional type (if optional itself doesn’t conform to a required protocol). This enables calls to properties and optional chaining of member chains starting from protocol metatypes.

Reviewer: ```suggestion
> **Note:** If a member returns a function type or an optional value, the type-checker considers “result type” (for the purposes of base type inference) to be a result type of a function type and/or wrapped value of an optional type (if `Optional` itself doesn’t conform to a required protocol). This enables calls to properties, and optional chaining of member chains, starting from protocol metatypes.
```

This approach works well for references without an explicit base, let’s consider an example:

```swift
protocol Style {
}

struct BorderedStyle : Style {
}

extension Style {
  static var bordered: BorderedStyle { BorderedStyle() }
}

func applyStyle<S: Style>(_: S) {
  // ...
}

applyStyle(.bordered)
```

> In this case (`applyStyle(.bordered)`) the reference to the member `.bordered` is re-written to be `Bordered.bordered` in the type-checked AST.

Reviewer: ```suggestion
In this case (`applyStyle(.bordered)`) the reference to the member `.bordered` is re-written to be `BorderedStyle.bordered` in the type-checked AST.
```

To make this work the type-checker would attempt to infer protocol conformance requirements from context, e.g. the call site of a generic function (in this case there is only one such requirement - the protocol `Style`), and propagate them to the type variable representing the implicit base type of the chain. If there is no other contextual information available, e.g. the result type couldn’t be inferred to some concrete type, the type-checker would attempt to bind base to the type of the inferred protocol requirement. 

> Member lookup filtering is adjusted to find static members on protocol metatype base but the `Self` part of the reference type is replaced with result type of the discovered member (looking through function types and IUOs) and additional conformance requirements are placed on it (the new `Self` type) to make sure that the new base does conform to the expected protocol.

Reviewer: ```suggestion
Member lookup filtering is adjusted to find static members on a protocol metatype base, but the `Self` part of the reference type is replaced with the result type of the discovered member (looking through function types and IUOs) and additional conformance requirements are placed on it (the new `Self` type) to make sure that the new base does conform to the expected protocol.
```

```swift
Toggle("Wi-Fi", isOn: $isWiFiEnabled)
  .toggleStyle(SwitchToggleStyle())
```

However, this approach has a few downsides:

* **Repetitive:** Only the “Switch” component of the style name is important, since we already know that the modifier expects a type of `ToggleStyle`.
* **Poor discoverability:** There is no autocomplete support to expose the available `ToggleStyle` types to choose from, so you have to know them in advance.

These downsides are impossible to avoid for generic parameters like above, which discourages generalizing functions. API designers should not have to choose between good design and easy-to-read code.

Instead, we could ideally support leading dot syntax for generic types with known protocol conformances, allowing syntax like this:

```swift
Toggle("Wi-Fi", isOn: $isWiFiEnabled)
  .toggleStyle(.switch)
```

### Existing Workarounds

There are ways of achieving the desired syntax today without changing the language, however they are often too complex and too confusing for API clients.

When SwiftUI was still in beta, it included one such workaround in the form of the `StaticMember` type:

```swift

> protocol ToggleStyle {
>   // ...
>   typealias Member = StaticMember<Self>
> }
> 
> struct StaticMember<Base> {
>   var base: Base
>   init(_ base: Base)
> }
> 
> extension StaticMember where Base: ToggleStyle {
>   static var switch: StaticMember<SwitchToggleStyle> { .init(.init()) }
>   static var checkbox: StaticMember<CheckboxToggleStyle> { .init(.init()) }
> }
> 
> extension View {
>   func toggleStyle<S : ToggleStyle>(_ style: S.Member) -> some View
> }
> 
> MyView().toggleStyle(.switch)
> MyView().toggleStyle(.checkbox)

Reviewer: - Use backticks with keyword properties
- Use same `Toggle("Wi-Fi", isOn: $isWiFiEnabled)` example?

```suggestion
// Rejected SwiftUI APIs:

public protocol ToggleStyle {
  // ...
  typealias Member = StaticMember<Self>
}

extension View {
  public func toggleStyle<S: ToggleStyle>(_ style: S.Member) -> some View
}

public struct StaticMember<Base> {
  public var base: Base
  public init(_ base: Base)
}

extension StaticMember where Base: ToggleStyle {
  public static var `default`: StaticMember<DefaultToggleStyle> { get }
  public static var `switch`: StaticMember<SwitchToggleStyle> { get }
  public static var checkbox: StaticMember<CheckboxToggleStyle> { get }
}

// Leading dot syntax (using rejected workaround):

Toggle("Wi-Fi", isOn: $isWiFiEnabled)
  .toggleStyle(.switch)
```

* Require that the result type of a member declaration conforms to the declaring protocol.

The second option is a much better choice that avoids having to do a global lookup and conformance checking and is consistent with semantics of leading dot syntax, namely, the requirement that result and base types of the chain have to be equivalent. This leads to a new rule: if the result type of a static member conforms to the declaring protocol, it should be possible to reference such a member on a protocol metatype, using leading dot syntax, by implicitly replacing the protocol with a conforming type.

> **Note:** If a member returns a function type or an optional value, the type-checker considers “result type” (for the purposes of base type inference) to be a result type of a function type and/or wrapped value of an optional type (if `Optional` itself doesn’t conform to a required protocol). This enables calls to properties, and optional chaining of member chains, starting from protocol metatypes.

This approach works well for references without an explicit base, let’s consider an example:

```swift

> protocol ToggleStyle {
> }
> 
> struct SwitchToggleStyle: ToggleStyle {
> }
> 
> extension ToggleStyle {
>   static var switch: SwitchToggleStyle { SwitchToggleStyle() }
> }
> 
> func applyStyle<S: ToggleStyle>(_: S) {
>   // ...
> }
> 
> applyStyle(.switch)

Reviewer: - Use backticks with keyword properties
- Use same `Toggle("Wi-Fi", isOn: $isWiFiEnabled)` example?

```suggestion
// Existing SwiftUI APIs:

public protocol ToggleStyle { ... }

public struct DefaultToggleStyle: ToggleStyle { ... }
public struct SwitchToggleStyle: ToggleStyle { ... }
public struct CheckboxToggleStyle: ToggleStyle { ... }

extension View {
  public func toggleStyle<S: ToggleStyle>(_ style: S) -> some View
}

// Possible SwiftUI APIs:

extension ToggleStyle {
  public static var `default`: DefaultToggleStyle { get }
  public static var `switch`: SwitchToggleStyle { get }
  public static var checkbox: CheckboxToggleStyle { get }
}

// Leading dot syntax (using proposed solution):

Toggle("Wi-Fi", isOn: $isWiFiEnabled)
  .toggleStyle(.switch)
```

```

> In this case (`applyStyle(.switch)`) the reference to the member `.switch` is re-written to be `SwitchToggleStyle.switch` in the type-checked AST.

Reviewer: ```suggestion
In the case of `.toggleStyle(.switch)`, the reference to the member `.switch` is re-written to be `SwitchToggleStyle.switch` in the type-checked AST.
```
